### PR TITLE
Upload screenshots and pages on failed builds

### PR DESCRIPTION
**Pull request by vassyz**
Because our flaky system tests are quite difficult to debug I looked into uploading a screenshot when the build fails.

I failed a random system test intentionally and the screenshots can be downloaded from the [artifacts link](https://github.com/DFE-Digital/apply-for-teacher-training/actions/runs/7050317739#artifacts).

Locally you can access the screenshot by clicking on the link in the console:
Screenshot 2023-11-30 at 18 16 00

I can disable this locally if people don't find it useful.

Let me know your thoughts @DFE-Digital/find-and-apply-devs .

@saliceti Can you please review the build code and let me know if it looks alright to you or if it needs improvements? Are there any concerns about storing files on GitHub? I presume they will be deleted once the PR is closed.

**Update:** I have also added a page save feature. This will help debugging by enabling the inspection of HTML code, as the majority of failures come from selector issues.

### CODE REVIEW

**Review by avinhurry — COMMENTED** [review]
What should the screenshot be of? I can see 6 [here](https://github.com/DFE-Digital/apply-for-teacher-training/actions/runs/7050317739#artifacts) and they are all of the application details page 🤔

**Comment by vassyz** [comment]
> What should the screenshot be of? I can see 6 [here](https://github.com/DFE-Digital/apply-for-teacher-training/actions/runs/7050317739#artifacts) and they are all of the application details page 🤔

System spec fails, a screenshot gets saved. What screenshots were you expecting?

It saves more files because tests run in parallel. I guess I could remove the timestamp and then they get overwritten. Even if multiple failures happen on the same line in a system spec, one screenshot should be enough.

**Comment by vassyz** [comment]
@avinhurry updated so now you'll only get one screenshot per failure.

**Comment by vassyz** [comment]
> It would be nice if there was a hosting feature where we could click a link and see the image in the browser directly from GH Actions instead of downloading a zip.
> 
> Is that doable?

The links will only be available after the build finishes, so they are not yet accessible. I don't want to invest too much time in this feature at this stage, as it was not explicitly requested. I wanted to see if it makes debugging flakey tests a bit easier.

**Comment by saliceti** [comment]
I don't thing artifacts would be deleted whenthe PR closes. You would have to include https://github.com/marketplace/actions/delete-artifact in the delete review app workflow
Also I would suggest setting a retention period anyway to make sure they are deleted anyway: https://github.com/actions/upload-artifact#retention-period

**Comment by vassyz** [comment]
@saliceti thanks for the suggestions! I added a retention period of 5 days and the delete step when removing the review app.

File: .github/workflows/delete-v2-review-app.yml [hunk 1]
```diff
@@ -106,3 +106,7 @@ jobs:
 
         env:
           DEPLOY_ENV: ${{env.DEPLOY_ENV}}
+      - uses: geekyeggo/delete-artifact@v2
+        with:
+          name: failure-uploads
```
Comment (saliceti) [inline .github/workflows/delete-v2-review-app.yml:111]: Don't we need to add the path as in https://github.com/DFE-Digital/apply-for-teacher-training/pull/8848/files#diff-ecad2183af1362d9d99f4a33e7491c1970c92255c3c318e688ece549ec91fe33R266 ?

Comment (vassyz) [reply .github/workflows/delete-v2-review-app.yml:111]: Isn't that the path it takes the files from? I'll test on this PR once it gets merged.

**Comment by vassyz** [comment]
@saliceti Can you please take another look when you're free? Thanks.

File: .github/workflows/delete-v2-review-app.yml (hunk 1 above)
Comment (saliceti) [reply .github/workflows/delete-v2-review-app.yml:111]: but the path is "path: /app/tmp/failures/" ?

**Review by JR-G — APPROVED** [review]

**Pull request merged**